stan/math/prim/functor/reduce_sum.hpp

```diff
@@ -20,76 +20,113 @@ namespace math {
+ * @tparam Args Parameter pack holding the types of the object send to
+ * `ReduceFunction::operator()`
+ */
 template <typename ReduceFunction, typename Enable, typename ReturnType,
           typename Vec, typename... Args>
 struct reduce_sum_impl {
-  return_type_t<Vec, Args...> operator()(const Vec& vmapped,
-                                         std::size_t grainsize,
-                                         std::ostream* msgs,
-                                         const Args&... args) const {
+  template <typename OpVec, typename... OpArgs,
+            typename = require_same_t<Vec, OpVec>,
+            typename = require_all_t<std::is_same<Args, OpArgs>...>>
+  return_type_t<OpVec, OpArgs...> operator()(OpVec&& vmapped,
+                                             std::size_t grainsize,
+                                             std::ostream* msgs,
+                                             OpArgs&&... args) const {
     const std::size_t num_jobs = vmapped.size();
 
-    if (num_jobs == 0)
+    if (num_jobs == 0) {
       return 0.0;
+    }
 
-    return ReduceFunction()(0, vmapped.size() - 1, vmapped, msgs, args...);
+    return ReduceFunction()(0, vmapped.size() - 1, std::forward<OpVec>(vmapped),
+                            msgs, std::forward<OpArgs>(args)...);
   }
 };
 
+/**
+ * Arithmetic definition of implimentation called by reduce `reduce_sum`.
+ * @tparam ReduceFunction An type with a valid `operator()`
+ * @tparam ReturnType The return type of the call to
+ * `ReduceFunction::operator()`
+ * @tparam Vec type of the first input argument with an `operator[]`
+ * @tparam Args Parameter pack holding the types of the object send to
+ * `ReduceFunction::operator()`
+ */
 template <typename ReduceFunction, typename ReturnType, typename Vec,
           typename... Args>
 struct reduce_sum_impl<ReduceFunction, require_arithmetic_t<ReturnType>,
                        ReturnType, Vec, Args...> {
+  /**
+   * Internal object used in `tbb::parallel_reduce`
+   * @note see link [here](https://tinyurl.com/vp7xw2t) for requirements.
+   */
   struct recursive_reducer {
-    using vmapped_t = Vec;
-    const vmapped_t& vmapped_;
+    Vec vmapped_;
     std::ostream* msgs_;
-    std::tuple<const Args&...> args_tuple_;
-    return_type_t<Vec, Args...> sum_;
+    std::tuple<Args...> args_tuple_;
+    return_type_t<Vec, Args...> sum_{0.0};
 
-    recursive_reducer(const vmapped_t& vmapped, std::ostream* msgs,
-                      const Args&... args)
-        : vmapped_(vmapped), msgs_(msgs), args_tuple_(args...), sum_(0.0) {}
+    template <typename OpVec, typename... OpArgs,
+              typename = require_same_t<Vec, OpVec>,
+              typename = require_all_t<std::is_same<Args, OpArgs>...>>
+    recursive_reducer(OpVec&& vmapped, std::ostream* msgs, OpArgs&&... args)
+        : vmapped_(std::forward<OpVec>(vmapped)),
+          msgs_(msgs),
+          args_tuple_(std::forward<OpArgs>(args)...) {}
```

Comment:
So with the forwarding and the && notation, if we pass in a reference, we'll get a tuple with a reference (not a copy of the object)?

Reply:
If we pass in an lvalue reference (a single `&` like `std::vector<T>&` or `const std::vector<T>&`) it will pass a normal reference to the args. If we pass in an rvalue ( `std::vector<T>&&`) then it knows the ownership of the underlying data for the vector can be moved from the input argument to the class's member. The link below is useful for this concept

https://isocpp.org/blog/2012/11/universal-references-in-c11-scott-meyers

Reply:
But I do agree the whole class could be simplified. Maybe it could be a lambda?

Reply:
I think it'll need to be a function. There's the generic version in prim (no threading, full high order autodiff support), a version in prim that supports only arithmetic types + is threaded, and then a version in rev specialized for vars.

I'm gonna leave it as is this pass through. After I push stuff tonight can you take a look at it? I think the type logic would be cleaner as a function as well.